### PR TITLE
[Dynamo] non-linear operator support + tests 

### PR DESCRIPTION
**Pull request by AndreSlavescu**
# [Dynamo] non-linear operator support + tests 

- Non-linear operator support from torch.nn.functional and torch.nn

## features

- added abs operator in fp32 and fp64 (cpu + cuda)
- tests for each torch.nn.functional and torch.nn operator

### CODE REVIEW

File: python/hidet/graph/ops/definitions/activation.py [hunk 1]
```diff
@@ -87,6 +88,131 @@ def __init__(self, x: Tensor):
...
+        super().__init__(
+            inputs=[x], task=SoftmaxTask(input_like(x, 'x'), axis), attributes={'axis': axis}, name='softmax'
+        )
+
+
+class SoftminOp(Operator):
+    def __init__(self, x: Tensor, axis: int = 1):
+        axis = normalize_dim(axis, len(x.shape))
+        super().__init__(
+            inputs=[-x], task=SoftmaxTask(input_like(x, 'x'), axis), attributes={'axis': axis}, name='softmin'
```
Comment (yaoyaoding) [inline python/hidet/graph/ops/definitions/activation.py:212]: It is better to do the minus in the Task instead of Operator definition.

Comment (yaoyaoding) [reply python/hidet/graph/ops/definitions/activation.py:212]: It's better to move the softmax task definition to activitation.py as well.

Comment (yaoyaoding) [reply python/hidet/graph/ops/definitions/activation.py:212]: Another option to deal with the `-` in softmin is to do this in the `def softmin()` function.

File: tests/operators/test_activation.py [hunk 2]
```diff
@@ -12,11 +12,16 @@
 import numpy as np
 import pytest
 import torch.nn.functional as F
+import torch.nn.functional as F
```
Comment (yaoyaoding) [inline tests/operators/test_activation.py:15]: duplicated?

File: tests/operators/test_activation.py [hunk 3]
```diff
@@ -12,11 +12,16 @@
...
 
 from hidet import ops
 from hidet.testing import check_unary, check_binary, check_torch_binary, check_torch_unary
 
 
+# hidet operators tested against numpy equivalent operators
+from hidet.testing import check_unary, check_binary, check_torch_binary, check_torch_unary
+
+
```
Comment (yaoyaoding) [inline tests/operators/test_activation.py:24]: duplicated?

Comment (yaoyaoding) [reply tests/operators/test_activation.py:24]: It might help to start from the lastest main branch every time when we start a new work, should avoid many conflicts like this.

File: python/hidet/graph/ops/definitions/activation.py [hunk 4]
```diff
@@ -87,6 +88,131 @@ def __init__(self, x: Tensor):
...
+class RReluOp(UnaryElementwiseOp):
+    def __init__(self, x: Tensor, lower: float = 1 / 8, upper: float = 1 / 3) -> Tensor:
+        super().__init__(
+            x, op=lambda v: if_then_else(v >= 0, v, prim.random(x.dtype(lower), x.dtype(upper)) * v), name='rrelu'
+        )
+
+
+class LogSigmoidOp(UnaryElementwiseOp):
+    def __init__(self, x: Tensor):
+        super().__init__(x, op=lambda v: prim.log(x.dtype(1.0) / (x.dtype(1.0) + prim.exp(-v))), name='logsigmoid')
```
Comment (yaoyaoding) [inline python/hidet/graph/ops/definitions/activation.py:148]: consider to simplify to -log(1+exp(-x)).

File: python/hidet/graph/ops/definitions/activation.py [hunk 5]
```diff
@@ -87,6 +88,131 @@ def __init__(self, x: Tensor):
...
+            op=lambda v: prim.max(x.dtype(0.0), v)
+            + prim.min(x.dtype(0.0), x.dtype(alpha) * (prim.exp(v / x.dtype(alpha)) - x.dtype(1.0))),
+            name='celu',
+        )
+
+
+class RReluOp(UnaryElementwiseOp):
+    def __init__(self, x: Tensor, lower: float = 1 / 8, upper: float = 1 / 3) -> Tensor:
+        super().__init__(
+            x, op=lambda v: if_then_else(v >= 0, v, prim.random(x.dtype(lower), x.dtype(upper)) * v), name='rrelu'
```
Comment (yaoyaoding) [inline python/hidet/graph/ops/definitions/activation.py:142]: might skip RReluOp as we do not have `prim.random` for now.

File: python/hidet/graph/ops/definitions/activation.py [hunk 6]
```diff
@@ -87,6 +88,131 @@ def __init__(self, x: Tensor):
...
+    def __init__(self, x: Tensor, threshold_val: float, value: float) -> Tensor:
+        super().__init__(x, op=lambda v: if_then_else(v > x.dtype(threshold_val), v, x.dtype(value)), name='threshold')
+
+
+class HardTanhOp(UnaryElementwiseOp):
+    def __init__(self, x: Tensor, min_val: float = -1.0, max_val: float = 1.0) -> Tensor:
+        super().__init__(
+            x,
+            op=lambda v: if_then_else(
+                v > x.dtype(max_val), x.dtype(max_val), if_then_else(v < x.dtype(min_val), x.dtype(min_val), v)
```
Comment (yaoyaoding) [inline python/hidet/graph/ops/definitions/activation.py:101]: consider to use the min/max primitive to implement this:
``out = max(min(v, min_val), max_val)`,
as it will only read v once.

File: python/hidet/graph/ops/definitions/activation.py [hunk 7]
```diff
@@ -87,6 +88,131 @@ def __init__(self, x: Tensor):
...
+        super().__init__(x, op=lambda v: v / (x.dtype(1.0) + prim.abs(v)), name='softsign')
+
+
+class SoftPlusOp(UnaryElementwiseOp):
+    def __init__(self, x: Tensor, beta: int = 1, threshold_val: int = 20):
+        super().__init__(
+            x,
+            op=lambda v: if_then_else(
+                v * x.dtype(beta) <= x.dtype(threshold_val),
+                (x.dtype(1.0) / x.dtype(beta)) * prim.log(x.dtype(1.0) + prim.exp(x.dtype(beta) * v)),
```
Comment (yaoyaoding) [inline python/hidet/graph/ops/definitions/activation.py:180]: better to do the 1/beta in python like x.dtype(1.0 / beta).

File: python/hidet/graph/ops/definitions/activation.py (hunk 6 above)
Comment (AndreSlavescu) [reply python/hidet/graph/ops/definitions/activation.py:101]: refactored to:

min(max_val, max(minval, v))

**Review by yaoyaoding — APPROVED** [review]
LGTM, thanks @AndreSlavescu.
